### ISSUE TITLE
The shutdown process does not proceed when a uaa token issuance failure occurs.

### ISSUE THREAD

**Issue by kukgini** [bosh-prometheus/firehose_exporter#28]
I am using prometheus v19.0.0 version. Therefore, firehose_exporter v4.2.5 is installed.

I do not have any CF related metrics recently, so I checked the log of firehose_exporter.

```
time="2018-03-29T07:04:25Z" level=info msg="Starting firehose_exporter (version=4.2.5, branch=master, revision=0d2a1ad41b0e47195aa39846ea3741592f8ac57d)" source="firehose_exporter.go:250"
time="2018-03-29T07:04:25Z" level=info msg="Build context (go=go1.9.2, user=root@092c0c1307bc, date=20171027-16:10:13)" source="firehose_exporter.go:251"
time="2018-03-29T07:04:25Z" level=info msg="Listening on :9186" source="firehose_exporter.go:328"
time="2018-03-29T07:04:25Z" level=info msg="Starting Firehose Nozzle..." source="firehose_nozzle.go:58"
time="2018-03-29T07:04:32Z" level=error msg="Error getting oauth token: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:04:32Z" level=error msg="Error while reading from the Firehose: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused" source="firehose_nozzle.go:111"
time="2018-03-29T07:04:34Z" level=error msg="Error getting oauth token: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:04:34Z" level=error msg="Error while reading from the Firehose: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused" source="firehose_nozzle.go:111"
time="2018-03-29T07:04:36Z" level=error msg="Error getting oauth token: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:04:36Z" level=error msg="Error while reading from the Firehose: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused" source="firehose_nozzle.go:111"
time="2018-03-29T07:04:40Z" level=error msg="Error getting oauth token: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:04:40Z" level=error msg="Error while reading from the Firehose: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused" source="firehose_nozzle.go:111"
time="2018-03-29T07:04:48Z" level=error msg="Error getting oauth token: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:04:48Z" level=error msg="Error while reading from the Firehose: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused" source="firehose_nozzle.go:111"
time="2018-03-29T07:05:04Z" level=error msg="Error getting oauth token: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:05:04Z" level=error msg="Error while reading from the Firehose: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused" source="firehose_nozzle.go:111"
time="2018-03-29T07:05:36Z" level=error msg="Error getting oauth token: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:05:36Z" level=error msg="Error while reading from the Firehose: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused" source="firehose_nozzle.go:111"
time="2018-03-29T07:06:36Z" level=error msg="Error getting oauth token: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:06:36Z" level=error msg="Error while reading from the Firehose: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused" source="firehose_nozzle.go:111"
time="2018-03-29T07:07:36Z" level=error msg="Error getting oauth token: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:07:36Z" level=error msg="Error while reading from the Firehose: Post https://uaa.mydomain.com/oauth/token: dial tcp <my.uaa.ip>:443: getsockopt: connection refused" source="firehose_nozzle.go:111"
time="2018-03-29T07:08:36Z" level=error msg="Error getting oauth token: Received a status code 503 Service Unavailable. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:08:36Z" level=error msg="Error while reading from the Firehose: Received a status code 503 Service Unavailable" source="firehose_nozzle.go:111"
time="2018-03-29T07:09:36Z" level=error msg="Error getting oauth token: Received a status code 503 Service Unavailable. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:09:36Z" level=error msg="Error while reading from the Firehose: Received a status code 503 Service Unavailable" source="firehose_nozzle.go:111"
time="2018-03-29T07:10:36Z" level=error msg="Error getting oauth token: Received a status code 503 Service Unavailable. Please check your Client ID and Secret." source="uaa_token_refresher.go:39"
time="2018-03-29T07:10:36Z" level=error msg="Error while reading from the Firehose: Received a status code 503 Service Unavailable" source="firehose_nozzle.go:111"
```

When I read the source code, it seems that the process shutdown should proceed after a few retries, but the shutdown did not actually proceed after three retries.

After restarting firehose_exporter with monit, the token issue has been resolved and is no longer reproducible.

Apart from the cause of failure to issue token in uaa, I suspect that the shutdown process is not executed after exceeding the number of retries. What do you think?

**Comment by frodenas** [comment]
@kukgini In case of failure, the default is to retry [1000](https://github.com/cloudfoundry/noaa/blob/master/consumer/async.go#L23) times. If you want to increase/decrease this value, you need to update the [firehose_exporter.doppler.max_retry_count](https://github.com/bosh-prometheus/prometheus-boshrelease/blob/master/jobs/firehose_exporter/spec#L29) property at the prometheus bosh release.

**Comment by kukgini** [comment]
@frodenas Interesting. Then I wonder why it did not retry more. The log finished at 2018-03-29. , I found this problem yesterday: 2018-04-6 (after 7 days). Is not it strange?

**Comment by frodenas** [comment]
There were some use cases were the `nooa` library close the channels on a retryalable error, leading to the exporter to read empty values indefinitely rather than catch the actual error (see #27). This should be fixed in [v5.0.2](https://github.com/bosh-prometheus/firehose_exporter/releases/tag/v5.0.2).

**Comment by frodenas** [comment]
Closing this issue. Please reopen if the problem still persists with [v5.0.2](https://github.com/bosh-prometheus/firehose_exporter/releases/tag/v5.0.2).
